modules/broker-core/src/main/java/org/wso2/broker/core/QueueHandler.java

```diff
@@ -124,14 +124,20 @@ Message dequeue() {
        
         Message message = queue.dequeue();
         if (message != null) {
-            pendingMessages.put(message.getMetadata().getInternalId(), message);
+            pendingMessages.put(message.getMetadata().getInternalId(), message.shallowCopy());
         }
         return message;
     }
 
     void acknowledge(long messageId) {
         // TODO handle nacks
-        pendingMessages.remove(messageId);
+        Message message = pendingMessages.remove(messageId);
+        message.release();
+    }
+
+    public void requeue(long messageId) {
+        Message message = pendingMessages.remove(messageId);
```

Comment:
What about the redelivered flag on the message?

Reply:
Fix done in 3fa16258b5751df6154b9a248ce577bbc09766ad